extend boostrap-setup to register extensions on prestart

Zetanova opened the pull request:
Currently there is only the possibility to add extensions prestart only over the config `akka.extensions`
and external initialized extensions are only possible to add after startup.

This PR fixes it with an `BootstrapSetup.Extensions` array.

Aaronontheweb commented on src/core/Akka/Actor/Internal/ActorSystemImpl.cs:
```diff
@@ -320,12 +322,21 @@ private void LoadExtensions()
+                .Select(_ => _.Extensions)
+                .GetOrElse(ImmutableArray<IExtensionId>.Empty);
+
+            ConfigureExtensions(extensions);
```
So `ConfigureExtensions` will be called twice if there are some extensions defined in HOCON _and_ some via the `BootstrapSetup`, correct? And the `RegisterExtension` method is idempotent, so even if the same extension was registered twice there shouldn't be any issues?

Aaronontheweb replied:
In the event that there's a HOCON collection for extensions and a `BootstrapSetup`, what happens?

Aaronontheweb reviewed (commented):
Looks good to me - but I did leave some questions on behavior and compatibility with the old HOCON

Aaronontheweb commented:
I think we ought to add a unit test for this

Zetanova commented:
I added new BootstrapSetupSpec and switched the DI Extension to BootstrapSetup.Extensions.
The DependencyResolverSetup and it's ExtensionIdProvider are having the same function.

Zetanova commented:
I am getting in a local test run a successful result,
but with the same error output

![image](https://user-images.githubusercontent.com/1385312/130314609-2e56aa22-2cba-44be-bce2-ff91e98dd15f.png)

Zetanova commented:
Why do this unit test builds fail?

Aaronontheweb commented:
> Why do this unit test builds fail?

Nothing to do with your PR - some of these specs are racy due to being designed to fit a bounded schedule, which is inherently brittle, and we haven't gone through and fixed all of them.

Aaronontheweb commented:
@Zetanova I will review this soon!

Aaronontheweb commented on src/contrib/dependencyinjection/Akka.DependencyInjection.Tests/DelegateInjectionSpecs.cs:
```diff
@@ -120,8 +120,9 @@ public AkkaService(IServiceProvider serviceProvider)
-                var setup = DependencyResolverSetup.Create(_serviceProvider)
-                    .And(BootstrapSetup.Create().WithConfig(TestKitBase.DefaultConfig));
+                var setup = BootstrapSetup.Create()
+                        .WithDependencyInjection(_serviceProvider)
```
Nice.

Aaronontheweb commented on src/contrib/dependencyinjection/Akka.DependencyInjection/DependencyResolver.cs:
```diff
@@ -111,15 +111,43 @@ public Props Props(Type type, params object[] args)
+        {
+        }
+
+        public DependencyResolverExtension(IDependencyResolver dependencyResolver)
```
You don't want this code on the `ExtensionIdProvider` - you want it on the `DependencyResolver` extension itself. The  `ExtensionIdProvider` is just meant to be a unique identifier that the `ActorSystem` uses to uniquely identify each possible extension to ensure there's at most one instance of it per-`ActorSystem`.

Aaronontheweb commented on src/contrib/dependencyinjection/Akka.DependencyInjection/DependencyResolverSetup.cs:
```diff
@@ -8,6 +8,23 @@
+    {
+        public static BootstrapSetup WithDependencyInjection(this BootstrapSetup bootstrap, IServiceProvider provider)
+        {
+            return bootstrap.WithExtensions(new DependencyResolverExtension(provider));
```
I'd just combine setups here and return an `ActorSystemSetup` - this is what we do with Phobos, for instance: https://phobos.petabridge.com/articles/setup/configuration.html#phobossetup-and-phobosconfigbuilder

```csharp
var phobosSetup = PhobosSetup.Create(phobosConfigBuilder)
    .And(BootstrapSetup.Create()
        .WithConfig(hocon)
        .WithActorRefProvider(PhobosProviderSelection.Cluster));
```

The `PhobosSetup.Create` method actually return an `ActorSystemSetup` because it combines both a `TracingSetup` and a separate `MonitoringSetup` type internally.

Aaronontheweb replied:
Using that type of composition is preferable to changing the `DependencyResolverExtension` - IMHO.

Aaronontheweb commented on src/core/Akka.Tests.Shared.Internals/AkkaSpec.cs:
```diff
@@ -67,6 +67,12 @@ public AkkaSpec(ActorSystemSetup setup, ITestOutputHelper output = null)
             BeforeAll();
         }
 
+        public AkkaSpec(Setup setup, ITestOutputHelper output = null)
```
Good idea

Aaronontheweb commented on src/core/Akka/Actor/ActorSystem.cs:
```diff
@@ -126,6 +133,8 @@ internal BootstrapSetup()
         /// </summary>
         public Option<ProviderSelection> ActorRefProvider { get; }
 
+        public ImmutableArray<IExtensionId> Extensions { get; }
```
Need to add an XML-DOC comment here

Zetanova replied:
If the ExtensionId (the extension factory instance) should not hold any state values / fields to setup the extensions,
would it not be better, to use just the Type? 
```csharp
public ImmutableArray<Type> Extensions { get; }
```

Aaronontheweb reviewed (changes requested):
See my suggestions around modifying the `DependencyResolverExtension` - otherwise this looks good.

Also - should add some tests to see what happens when extensions are defined both via HOCON and `BootstrapSetup`

Zetanova reviewed (commented):
> In the event that there's a HOCON collection for extensions and a BootstrapSetup, what happens?

The same as trying to add the same extension twice, first wins and the rest gets ignored.
The point of the new `BoostrapSetup.Extensions` list setup and register custom instances of `Extension` factories.

Zetanova reviewed (commented):
i didn't press the review button

Zetanova commented:
I looked now at the `ActorSystemImpl._extensions` its a mess, needs to be redone.

It registers Lazy Factory Functions one-by-one (foreach) and instantly initializing it. 

It is absolutely pointless.

The only thing what can happen is that the extension constructor can query itself
because of the safe-execution protection of the Lazy instance.
